File: Problems & Solutions/491. Non-decreasing Subsequences/20 January 2022/Approach 1 - Backtracking/Non-decreasing Subsequences.cpp
class Solution {
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        set<vector<int>> result;
        vector<int> sequence;
        function<void(int)> backtrack;
        backtrack = [&nums, &sequence, &result, &backtrack](int index) -> void {
            if (index == nums.size()) {
                if (sequence.size() >= 2) {
                    result.insert(sequence);
                }
                return;
            }
            if (sequence.empty() || sequence.back() <= nums[index]) {
                sequence.push_back(nums[index]);
                // call recursively
                backtrack(index + 1);
                sequence.pop_back();
            }
            backtrack(index + 1);
        };
        backtrack(0);
        return vector(result.begin(), result.end());
    }
};

File: Problems & Solutions/491. Non-decreasing Subsequences/20 January 2022/Approach 1 - Backtracking/Non-decreasing Subsequences.py
class Solution:
    def findSubsequences(self, nums: List[int]) -> List[List[int]]:
        result = set()
        sequence = []

        def backtrack(index):
            if index == len(nums):
                if len(sequence) >= 2:
                    result.add(tuple(sequence))
                return
            if not sequence or sequence[-1] <= nums[index]:
                sequence.append(nums[index])
                backtrack(index + 1)
                sequence.pop()
            backtrack(index + 1)
        backtrack(0)
        return result